### ISSUE TITLE
fix(moe): VKERNELS_MXFP4_BF16 produces degenerate serving output on K3/gfx942 — never passed a correctness gate

### ISSUE THREAD

**Issue by xzyaoi** [opentela-ai/vkernels#74]
## Summary

The `VKERNELS_MXFP4_BF16` fused-MoE backend (issue #20 integration,
`vkernels_experts.py` → `vk_hip_fused_moe_mxfp4`) produces **degenerate
serving output** on Kimi-K3 / gfx942 (MI300A). It has never passed a
correctness gate in serving; the "PASS" evidence attached to #20 was a
`GEN_CORRECTNESS_SMOKE=1` non-empty check, which cannot distinguish coherent
text from garbage.

## Evidence (Beverin, K3 serve TP8×PP3, real checkpoint)

Raw `capital -> Paris` continuations from `gen_correctness_*.json` —
independent of any verdict field:

| jobs | MoE backend | weights | continuation | real? |
|---|---|---|---|---|
| 588856, 589458 | `TRITON_UNFUSED` | `auto` real | `' Paris.",\n+  "The Eiffel Tower is locate'` | **coherent** |
| 589456 | `TRITON_UNFUSED` | `dummy` | `'.dartampionship.dartampionship…'` | negative control |
| 597880–603711 | `VKERNELS_MXFP4_BF16` | `auto` real | `'.dartampionship…'`; 597880 `'!!!!…'` | **degenerate** |
| 639143, 639260 | `VKERNELS_MXFP4_BF16` | `auto` real | `'!!!!…'` | **degenerate** |

- **589456 vs 589458** isolate the variable to the *weights*, not the
  pipeline: same `TRITON_UNFUSED` backend, `dummy` → garbage, real → coherent.
- **Every** real-weight run on `VKERNELS_MXFP4_BF16` is degenerate, so the
  backend — not the model, checkpoint or attention path — is at fault.
- `'!!!!…'` is token id 0 repeated: the sampler argmax-ing all-NaN logits
  (K3 tokenizer id 0 == `'!'`). So the failure spans two modes: wrong-but-
  finite (`.dartampionship…`) and NaN (`!!!!…`).
- Jobs 597880 and 603711 ran with `GEN_CORRECTNESS_SMOKE=1` (real weights,
  `load_format=auto`, `[SMOKE] probing` in the log). The smoke matcher is
  `_nonempty`, so `.dartampionship…` and `!!!!…` both "passed". That is why
  the defect went unnoticed.

## Immediate mitigation (landed)

`sitecustomize.py` now gates the `VKERNELS_MXFP4_BF16` entry of
`_get_priority_backends` behind `VKERNELS_MOE` (default `1`, unchanged);
`VKERNELS_MOE=0` falls through to the known-good `TRITON_UNFUSED` backend.
Job 639740 confirms `Using 'TRITON_UNFUSED' Mxfp4 MoE backend`. This
unblocks the issue #45 attention acceptance, which is otherwise confounded by
this defect.

## What needs doing

The standalone kernel path is validated in-tree
(`meta/benchmarks/test_moe_fused_bigshape_correct.hip` covers E up to 256,
ispp 512–33792 including the >2^31 expert-offset wrap, and SiTU decode/prefill
— all `PASS`), so the defect is most likely in the **serving integration**,
not the kernel maths. Candidates, in order:

1. **Weight layout / scales.** For `VKERNELS_MXFP4_BF16`,
   `convert_gpt_oss_weight_to_mxfp4_moe_kernel_format` is patched to pass
   weights and scales through unchanged (no `_swizzle_mxfp4`, no
   `PrecisionConfig`). If the checkpoint layout is not what
   `vk_hip_fused_moe_mxfp4` expects, the experts read garbage — finite but
   wrong (`.dartampionship…`) or NaN.
2. **Routing metadata.** `sids`/`eids` construction, `topk` padding and the
   `EM` high-water mark (`vkernels_experts.py` `_get_persistent_*`).
3. **Activation.** SiTU / `activation_situ_linear_beta` handling on the
   serving path (the standalone test covers SiTU, so lower priority).
4. **Shape-specific edge** at K3's real MoE shape (h=7168, ispp=3072,
   E=896, TP8 local E=112) — the standalone sweep brackets but does not
   include ispp=3072.

## Repro

`serve_kimi_k3_otela_beverin.sbatch` with `VKERNELS_MXFP4_BF16` selected
(i.e. `VKERNELS_MOE=1`, the default) and `GEN_PROBE=1` **without**
`GEN_CORRECTNESS_SMOKE=1`; inspect `gen_correctness_<job>.json` `results[].text`.
Adding a naive-reference comparison at the K3 serving shape (h=7168,
ispp=3072, E=112) would give a serving-independent oracle for the layout
question.

Full write-up:
`meta/benchmarks/artifacts/k3-attn-accept/README.md` (ERRATUM section).

**Comment by xzyaoi** [comment]
## Root cause — serving weight-format contract (fixed)

The issue's candidate #1 was right about the mechanism but the detail differs: the conversion was never "patched to pass through" — **no such patch existed anywhere in the deploy stack** (`deploy*/k3_patch.py`, `deploy*/sitecustomize.py`, `home/pylib/*` reference neither `convert_weight_to_mxfp4_moe_kernel_format` nor `convert_gpt_oss_weight_to_mxfp4_moe_kernel_format`).

What actually happened: the #78 shim registers `VkernelFusedExperts` by hijacking `oracle.mxfp4.backend_to_kernel_cls` for **`Mxfp4MoeBackend.AITER_MXFP4_BF16`**. The selector therefore returns the AITER enum, and `Mxfp4MoEMethod._setup_kernel` (`quantization/mxfp4.py:723`) runs the oracle conversion's **AITER branch** — `aiter.ops.shuffle.shuffle_weight(is_guinterleave=True, gate_up=True)` + `shuffle_scale`, the **gfx950 CK shuffle layout**.

`vk_hip_fused_moe_mxfp4` consumes the **raw `create_weights()` layout** (w13 `[E, 2·ispp, hidden/2]` uint8, gate rows `[0, ispp)` / up rows `[ispp, 2·ispp)` separated halves; w2 `[E, hidden, ispp/2]`; N-major ue8m0 scales; `float*` biases — `hip_capi.cpp:40`). It got CK-shuffled weights *and* CK-shuffled scales → experts read garbage (finite-wrong `.dartampionship…`) or NaN (`!!!!…`, token 0) — exactly the two degenerate modes in the issue table — while `TRITON_UNFUSED` (oracle conversion matched to its own kernel) stayed coherent on the same checkpoint.

## Fix (branch `issue-74-mxfp4-serving-fix`, base `4c58c58`, pushed)

- `src/python/vkernels/vllm_experts.py`:
  - `convert_weights_for_vkernel()` — the serving contract: weights/scales pass through **unchanged**, bf16 biases cast to the C ABI's fp32; accepts both oracle call shapes.
  - `register_vkernel_backend_shim()` — the coupled registration the sitecustomize previously improvised: experts-class hijack **plus** interception of both oracle `convert_*` functions so the `AITER_MXFP4_BF16` enum routes to the vkernels pass-through.
- Deployment (`home/pylib/vkernels_experts.py` + `sitecustomize.py`) patched to call `install_vkernel_weight_conversion()` (backups `*.bak-issue74`, idempotent). Full write-up: `meta/benchmarks/artifacts/k3-mxfp4-serving-fix/README.md`.

## The correctness gate that never existed

`tests/python/test_mxfp4_serving_parity.py` — `vk_hip_fused_moe_mxfp4` vs a naive bf16 dequant reference (single bf16 rounding, fp32 accumulation, K3 SiTU, cancellation-aware tolerance mirroring `test_moe_fused_bigshape_correct.hip`), plus shim-integration tests against the real vLLM oracle, plus a negative control (pair-interleaved gate/up rows must FAIL parity).

**Evidence (beverin, Slurm `vk-i74-parity`, jobs 640417/640420):**

```
ConvertWeightsForVkernelTest        2/2 ok
VkernelBackendShimTest              3/3 ok   (AITER enum -> [VkernelFusedExperts];
                                              conversion returns raw uint8 + fp32 biases)
Mxfp4ServingParityTest              decode (M=8, bs16) ok · prefill (M=64, bs64) ok
  negative control (interleaved layout) ok — caught, as it must be
  K3 SERVING SHAPE h=7168, ispp=3072, E=112, topk=8  ok   ← the shape the
                                              bigshape sweep brackets but never covered
```

## Remaining risk

The end-to-end K3 serve (`capital -> Paris`, real weights, `VKERNELS_MXFP4_BF16`) has not been re-run; the fix changes only the weight conversion, which is now covered serving-independently at the exact serving shape. Next scheduled serve should confirm via `gen_correctness_*.json` (and the smoke matcher now records `smoke: true/false`, so a non-empty-only "pass" can't recur).

**Comment by xzyaoi** [comment]
Fixed via #120 (merged): serving weight-format contract. convert_weights_for_vkernel now passes raw uint8 weights + fp32 biases to vk_hip_fused_moe_mxfp4, and the shim intercepts both oracle convert_* entry points so the AITER_MXFP4_BF16 enum routes to the vkernels pass-through. Parity verified at the K3 serving shape (h=7168, ispp=3072, E=112). The end-to-end serve later confirmed coherent output with the shim + issue-74 conversion installed (PR #104, job 640737, verdict=PASS). Closing.
